### PR TITLE
feat: Added `--force-install` CLI flag for meltano `add` command.

### LINKED ISSUE #7277: feature: `--force` CLI flag for `meltano add`
### Feature scope

CLI (options, error messages, logging, etc.)

### Description

We currently support the `--force` CLI flag for `meltano install`. Since `meltano add` installs by default, it should also support this flag.

### PR DESCRIPTION
**Pull request by XshubhamX**
**Additions:**

1. Added the `--force` flag support for `add` command.
2. Added test.
3. Updated the doc for the same

cc @WillDaSilva 

Closes https://github.com/meltano/meltano/issues/7277

### CODE REVIEW

File: docs/docs/reference/command-line-interface.md [hunk 1]
```diff
@@ -59,6 +59,8 @@ meltano add <type> <name>
 
 # For example:
 meltano add extractor tap-gitlab
+# Or you can also add a package forcefully using --force flag
```
Comment (WillDaSilva) [inline docs/docs/reference/command-line-interface.md:62]:
```suggestion
# You can ignore the required Python declared by a plugin by using --force flag
```

File: docs/docs/reference/command-line-interface.md [hunk 2]
```diff
@@ -67,6 +69,8 @@ Without a `--custom`, `--inherit-from` or `--from-ref` option, this will add the
 to your [`meltano.yml` project file](/concepts/project#plugins)
 using a [shadowing plugin definition](/concepts/project#shadowing-plugin-definitions).
 
+The `--force` will ignore the required Python version declared by the plugins.
```
Comment (edgarrmondragon) [inline docs/docs/reference/command-line-interface.md:72]: Can you move this to the `Parameters` section of this command below?

File: tests/meltano/cli/test_add.py [hunk 3]
```diff
@@ -793,3 +793,16 @@ def test_add_with_python_version(self, cli_runner: CliRunner):
                 ),
             )
             venv_mock.assert_called_once_with(python)
+
+    def test_add_with_force_flag(self, project: Project, cli_runner: CliRunner):
+        with mock.patch("meltano.cli.add.install_plugins") as install_plugin_mock:
+            install_plugin_mock.return_value = True
+            cli_runner.invoke(
+                cli, ["add", "transform", "tap-google-analytics", "--force"]
```
Comment (edgarrmondragon) [inline tests/meltano/cli/test_add.py:801]:
```suggestion
                cli,
                ["add", "transform", "tap-google-analytics", "--force"],
```

File: docs/docs/reference/command-line-interface.md [hunk 4]
```diff
@@ -59,6 +59,8 @@ meltano add <type> <name>
 
 # For example:
 meltano add extractor tap-gitlab
+# Or you can also add a package forcefully using --force flag
+meltano add extractor tap-gitlab --force
```
Comment (edgarrmondragon) [inline docs/docs/reference/command-line-interface.md:63]:
```suggestion
meltano add extractor tap-gitlab --force
```

File: docs/docs/reference/command-line-interface.md
Comment (edgarrmondragon) [inline docs/docs/reference/command-line-interface.md:1]: Let's add a similar section to what we have for `install` above the `Note: Unlike...` note:

> If the plugin you are trying to add declares that it does not support the version of Python you are using, but you want to attempt to use it anyway, you can override the Python version restriction by providing the `--force` flag to `meltano add`.

**Review by edgarrmondragon — CHANGES_REQUESTED** [review]
Thanks @XshubhamX! Requested a few changes.

**Comment by edgarrmondragon** [comment]
You might wanna update instances of `install_plugin_mock.assert_called_*` to check the `force` parameter you're now passing it.

**Comment by ReubenFrankel** [comment]
Great timing! :smile: We discussed this exact feature recently here: https://github.com/meltano/meltano/pull/7907#issuecomment-1708096810

Will this work as described in #8110?

> As discussed in #7907, a `--force` (or similar) flag would enable a plugin to be re-added to the project, regardless of whether or not the plugin already exists. This would essentially be a shortcut for:
>
> ```
> meltano remove <plugin type> <plugin name>
> meltano add <plugin type> <plugin name>
> ```
>
> This feature would be useful in resetting a plugin definition that perhaps isn't quite working after some user changes, updating a custom plugin that doesn't have a `.lock` file (i.e. added from `--custom` or [`--from-ref`](#7907)), or converting from a custom plugin to the Hub definition once published.
>
> Adding a new plugin with this flag should just `meltano add` as normal.

**Comment by edgarrmondragon** [comment]
> Great timing! 😄 We discussed this exact feature recently here: [#7907 (comment)](https://github.com/meltano/meltano/pull/7907#issuecomment-1708096810)
> 
> Will this work as described in #8110
> 
> > As discussed in #7907, a `--force` (or similar) flag would enable a plugin to be re-added to the project, regardless of whether or not the plugin already exists. This would essentially be a shortcut for:
> > ```
> > meltano remove <plugin type> <plugin name>
> > meltano add <plugin type> <plugin name>
> > ```
> > 
> > 
> >     
> >       
> >     
> > 
> >       
> >     
> > 
> >     
> >   
> > This feature would be useful in resetting a plugin definition that perhaps isn't quite working after some user changes, updating a custom plugin that doesn't have a `.lock` file (i.e. added from `--custom` or [`--from-ref`](#7907)), or converting from a custom plugin to the Hub definition once published.
> > Adding a new plugin with this flag should just `meltano add` as normal.
> 
> , or is this just a way to pass `--force` to the implicit `meltano install` step of `meltano add`?

Oh good call @ReubenFrankel. This is NOT https://github.com/meltano/meltano/issues/8110 😬.

I didn't realize we had two feature requests that suggested the same flag for different ends:

* https://github.com/meltano/meltano/issues/8110
* https://github.com/meltano/meltano/issues/7277

**Comment by XshubhamX** [comment]
@WillDaSilva @edgarrmondragon Please check the PR, I have made the required changes

**Comment by WillDaSilva** [comment]
@XshubhamX @edgarrmondragon @ReubenFrankel I think `--force` is fine here (if only to maintain consistency with `meltano install`), and I'd suggest we use `--reinstall` for #8110. I've left a comment on #8110 with my thoughts:
- https://github.com/meltano/meltano/issues/8110#issuecomment-1729845907

Please let me know what you all think of this approach.

**Comment by ReubenFrankel** [comment]
@edgarrmondragon @WillDaSilva I'm happy to be convinced otherwise, but I actually think the opposite. :sweat_smile: https://github.com/meltano/meltano/issues/8110#issuecomment-1729855158

**Comment by edgarrmondragon** [comment]
@XshubhamX Let's use `--force-install` instead of `--force`. We can get rid of the short option `-f`.

**Comment by XshubhamX** [comment]
> @XshubhamX With the context given [above](https://github.com/meltano/meltano/pull/8169#issuecomment-1729739870), what do you think of using a different flag name for this feature, e.g. `--force-install`. Alternatively, we could suggest a different flag for #8110. Not requesting changes at the moment, just trying to get your thoughts on it.
> 
> cc @WillDaSilva @ReubenFrankel

I also think that using --force-install flag instead of --force flag quite fits with the already existing --no-install flag implementation.

**Comment by edgarrmondragon** [comment]
@XshubhamX Given the additional context in https://github.com/meltano/meltano/issues/8110#issuecomment-1730124598, let's keep `--force` here.

Sorry for the back and forth 😅

**Comment by XshubhamX** [comment]
> @XshubhamX Let's use `--force-install` instead of `--force`. We can get rid of the short option `-f`.

@edgarrmondragon  We are going ahead with this, right?

**Comment by edgarrmondragon** [comment]
> > @XshubhamX Let's use `--force-install` instead of `--force`. We can get rid of the short option `-f`.
> 
> @edgarrmondragon We are going ahead with this, right?

Yeah, let's rename it :+1:

**Comment by XshubhamX** [comment]
@edgarrmondragon PR is updated!

**Review by edgarrmondragon — APPROVED** [review]
Thanks @XshubhamX!

**Pull request merged**